diff -ruN a/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.c b/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.c
--- a/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.c	2021-12-22 06:31:09.863228014 +0000
+++ b/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.c	2021-12-22 06:32:02.687811594 +0000
@@ -30,7 +30,7 @@
 static int ngx_http_lua_socket_tcp_close(lua_State *L);
 static int ngx_http_lua_socket_tcp_settimeout(lua_State *L);
 static int ngx_http_lua_socket_tcp_settimeouts(lua_State *L);
-static void ngx_http_lua_socket_tcp_handler(ngx_event_t *ev);
+void ngx_http_lua_socket_tcp_handler(ngx_event_t *ev);
 static ngx_int_t ngx_http_lua_socket_tcp_get_peer(ngx_peer_connection_t *pc,
     void *data);
 static void ngx_http_lua_socket_init_peer_connection_addr_text(
@@ -52,13 +52,13 @@
     ngx_http_lua_socket_tcp_upstream_t *u);
 static ngx_int_t ngx_http_lua_socket_test_connect(ngx_http_request_t *r,
     ngx_connection_t *c);
-static void ngx_http_lua_socket_handle_conn_error(ngx_http_request_t *r,
+void ngx_http_lua_socket_handle_conn_error(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, ngx_uint_t ft_type);
 static void ngx_http_lua_socket_handle_read_error(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, ngx_uint_t ft_type);
 static void ngx_http_lua_socket_handle_write_error(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, ngx_uint_t ft_type);
-static void ngx_http_lua_socket_handle_conn_success(ngx_http_request_t *r,
+void ngx_http_lua_socket_handle_conn_success(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u);
 static void ngx_http_lua_socket_handle_read_success(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u);
@@ -80,7 +80,7 @@
 static void ngx_http_lua_socket_resolve_handler(ngx_resolver_ctx_t *ctx);
 static int ngx_http_lua_socket_resolve_retval_handler(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, lua_State *L);
-static int ngx_http_lua_socket_conn_error_retval_handler(ngx_http_request_t *r,
+int ngx_http_lua_socket_conn_error_retval_handler(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, lua_State *L);
 static int ngx_http_lua_socket_read_error_retval_handler(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, lua_State *L);
@@ -141,7 +141,7 @@
     int socket_op);
 static void ngx_http_lua_tcp_queue_conn_op_cleanup(void *data);
 static void ngx_http_lua_tcp_resolve_cleanup(void *data);
-static void ngx_http_lua_coctx_cleanup(void *data);
+void ngx_http_lua_coctx_cleanup(void *data);
 static void ngx_http_lua_socket_free_pool(ngx_log_t *log,
     ngx_http_lua_socket_pool_t *spool);
 static int ngx_http_lua_socket_shutdown_pool(lua_State *L);
@@ -1545,7 +1545,7 @@
 }
 
 
-static int
+int
 ngx_http_lua_socket_conn_error_retval_handler(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, lua_State *L)
 {
@@ -2017,12 +2017,14 @@
     u_char           errstr[NGX_MAX_ERROR_STR];
     u_char          *p;
 
-    if (ft_type & (NGX_HTTP_LUA_SOCKET_FT_RESOLVER
-                   | NGX_HTTP_LUA_SOCKET_FT_SSL))
-    {
+    if (ft_type & NGX_HTTP_LUA_SOCKET_FT_RESOLVER) {
         return 2;
     }
 
+    if (ft_type & NGX_HTTP_LUA_SOCKET_FT_SSL) {
+        return 0;
+    }
+
     lua_pushnil(L);
 
     if (ft_type & NGX_HTTP_LUA_SOCKET_FT_TIMEOUT) {
@@ -3221,7 +3223,7 @@
 }
 
 
-static void
+void
 ngx_http_lua_socket_tcp_handler(ngx_event_t *ev)
 {
     ngx_connection_t                *c;
@@ -3424,7 +3426,7 @@
 }
 
 
-static void
+void
 ngx_http_lua_socket_handle_conn_success(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u)
 {
@@ -3536,7 +3538,7 @@
 }
 
 
-static void
+void
 ngx_http_lua_socket_handle_conn_error(ngx_http_request_t *r,
     ngx_http_lua_socket_tcp_upstream_t *u, ngx_uint_t ft_type)
 {
@@ -6091,7 +6093,7 @@
 }
 
 
-static void
+void
 ngx_http_lua_coctx_cleanup(void *data)
 {
     ngx_http_lua_socket_tcp_upstream_t      *u;
diff -ruN a/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.h b/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.h
--- a/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.h	2021-12-22 06:31:09.863228014 +0000
+++ b/ngx_lua-0.10.20/src/ngx_http_lua_socket_tcp.h	2021-12-22 06:32:58.468225974 +0000
@@ -120,6 +120,9 @@
 
 #if (NGX_HTTP_SSL)
     ngx_str_t                        ssl_name;
+    ngx_ssl_session_t               *ssl_session_ret;
+    const char                      *error_ret;
+    int                              openssl_error_code_ret;
 #endif
 
     unsigned                         ft_type:16;
