diff --git a/nginx-1.15.8/src/http/ngx_http_upstream.c b/nginx-1.15.8/src/http/ngx_http_upstream.c
index 90710557..539a4db9 100644
--- a/nginx-1.15.8/src/http/ngx_http_upstream.c
+++ b/nginx-1.15.8/src/http/ngx_http_upstream.c
@@ -8,6 +8,9 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_http.h>
+#if (NGX_HTTP_LUA_KONG)
+#include <ngx_http_lua_kong_module.h>
+#endif


 #if (NGX_HTTP_CACHE)
@@ -1719,6 +1722,10 @@ ngx_http_upstream_ssl_init_connection(ngx_http_request_t *r,
         }
     }

+#if (NGX_HTTP_LUA_KONG)
+    ngx_http_lua_kong_set_upstream_ssl(r, c);
+#endif
+
     r->connection->log->action = "SSL handshaking to upstream";

     rc = ngx_ssl_handshake(c);
